### PR TITLE
Save Tendermint's state to the DB

### PR DESCRIPTION
**Pull request by majecty**
Tendermint should save votes to the persistent DB not to send
DoubleVote. Save backup data when step is changed or the node sends a
vote.

### CODE REVIEW

File: core/src/consensus/tendermint/backup.rs [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+use kvdb::{DBTransaction, KeyValueDB};
+use primitives::H256;
+
+use super::message::ConsensusMessage;
+use super::types::{Height, Step, View};
+use crate::db;
+
+const BACKUP_KEY: &[u8] = b"tendermint-backup";
+
+pub struct BackupData<'a> {
```
Comment (sgkim126) [inline core/src/consensus/tendermint/backup.rs]: BackupView

File: core/src/consensus/tendermint/backup.rs [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    pub height: &'a Height,
+    pub view: &'a View,
+    pub step: &'a Step,
+    pub votes: &'a [ConsensusMessage],
+    pub last_seal: &'a (H256, View, Vec<Signature>),
+}
+
+pub struct BackupOut {
```
Comment (sgkim126) [inline core/src/consensus/tendermint/backup.rs]: BackupData

File: core/src/consensus/tendermint/backup.rs [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    pub height: Height,
+    pub view: View,
+    pub step: Step,
+    pub votes: Vec<ConsensusMessage>,
+    pub proposal: Option<H256>,
+    pub last_seal: (H256, View, Vec<Signature>),
+}
+
+pub struct Backup {
+    db: Arc<KeyValueDB>,
```
Comment (sgkim126) [inline core/src/consensus/tendermint/backup.rs]: I don't think we need to create `Backup` struct.
How about making `backup` and `load` free functions that take the reference of `KeyValueDB`?

Comment (sgkim126) [reply core/src/consensus/tendermint/backup.rs]: And I prefer `restore` to `load`.

Comment (majecty) [reply core/src/consensus/tendermint/backup.rs]: Good I fixed it.

File: core/src/consensus/tendermint/mod.rs [hunk 4]
```diff
@@ -454,6 +458,99 @@ impl Tendermint {
...
+        let c = self.client.read().as_ref().and_then(|weak| weak.upgrade()).expect("Client should be initialized");
+
+        let backup = restore(c.get_kvdb().as_ref());
+        if let Some(backup) = backup {
+            *self.step.write() = backup.step;
+            self.height.store(backup.height, AtomicOrdering::SeqCst);
+            self.view.store(backup.view, AtomicOrdering::SeqCst);
+            *self.last_seal.write() = backup.last_seal;
+            if let Some(proposal) = backup.proposal {
+                if let Some(_header) = c.block_header(&BlockId::Hash(proposal)) {
```
Comment (sgkim126) [inline core/src/consensus/tendermint/mod.rs]:
```rust
if c.block_header(&BlockId::Hash(proposal)).is_some() {
```

File: core/src/consensus/tendermint/mod.rs [hunk 5]
```diff
@@ -454,6 +458,99 @@ impl Tendermint {
...
+                    *self.proposal.write() = Some(proposal);
+                }
+            }
+            let prev_header =
+                c.block_header(&BlockId::Number(backup.height as u64 - 1)).expect("Prev block should be imported");
+            *self.proposal_parent.write() = prev_header.hash();
+
+            let mut to_next_height = None;
+            for vote in backup.votes {
```
Comment (sgkim126) [inline core/src/consensus/tendermint/mod.rs]: Should we do it for all votes? or we just need the last vote?
```rust
backup.votes.iter().rev().filter_map(|vote| self.load_vote_from_backup(&vote, backup.height)).take(1);
```

File: core/src/consensus/tendermint/mod.rs (hunk 4 above)
Comment (majecty) [reply core/src/consensus/tendermint/mod.rs]: I fixed it.

File: core/src/consensus/tendermint/backup.rs [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+        height,
+        view,
+        step,
+        votes,
+        last_confirmed_view,
+    } = backup_data;
+    let mut s = rlp::RlpStream::new();
+    s.begin_list(5);
+    s.append(height).append(view).append(step).append_list(votes);
+    s.begin_list(2).append(&last_confirmed_view.0).append(&last_confirmed_view.1);
```
Comment (sgkim126) [inline core/src/consensus/tendermint/backup.rs]: How about flatten this list?
```rust
s.begin_list(6);
s.append(height).append(view).append(step).append_list(votes);
s.append(&last_confirmed_view.0).append(&last_confirmed_view.1);
```

File: core/src/consensus/tendermint/backup.rs [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+        proposal,
+        last_confirmed_view,
+    })
+}
+
+fn find_proposal(votes: &[ConsensusMessage], height: Height, view: View) -> Option<H256> {
+    votes
+        .iter()
+        .rev()
+        .filter_map(|vote| {
```
Comment (sgkim126) [inline core/src/consensus/tendermint/backup.rs]:
```rust
    .find(|vote|  vote.vote_step.step == Step::Propose && vote.vote_step.view == view && vote.vote_step.height == height)
    .map(|vote| vote.block_hash)
});

File: core/src/consensus/tendermint/backup.rs (hunk 6 above)
Comment (majecty) [reply core/src/consensus/tendermint/backup.rs]: Good!

**Comment by majecty** [comment]
@sgkim126 I fixed them.

**Pull request merged**